File: Data Structures & Algorithms/binary-tree-diameter/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        result = [0]


        def calculateHeight(root):
            if not root:
                return 0
            
            result_left = calculateHeight(root.left)
            result_right = calculateHeight(root.right)

            result[0] = max(result[0], result_left + result_right)

            return 1 + max(result_left, result_right)
        
        calculateHeight(root)

        return result[0]

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        return countNode(root, root.val);
    }

    public int countNode(TreeNode root, int maxValue){
        if(root == null){
            return 0;
        }

        int res = (maxValue <= root.val) ? 1 : 0;
        maxValue = Math.max(root.val, maxValue);
        res += countNode(root.left, maxValue);
        res += countNode(root.right, maxValue);

        return res;
    }
}
